File: BinaryTree/Python/BinaryTree.py
# Full Binary Tree
# Perfect Binary Tree
# Complete Binary Tree

class Node:
    def __init__(self, data):
        self.data = data
        self.right = None
        self.left = None


def is_full_bt(node):
    if node is None:
        return True
    if node.left is None and node.right is None:
        return True
    if node.left is not None and node.right is not None:
        return is_full_bt(node.left) and is_full_bt(node.right)
    return False


def calculate_depth(node):
    d = -1  # Initialize
    while node is not None:
        d += 1
        node = node.left
    return d


def is_perfect_bt(node, d, level=0):  # Default level value is 0
    if node is None:
        return True
    if node.left is None and node.right is None:
        return d == level + 1
    if node.left is None or node.right is None:
        return False
    return is_perfect_bt(node.left, d, level + 1) and is_perfect_bt(node.right, d, level + 1)


def count_node(node):
    if node is None:
        return 0
    return 1 + count_node(node.left) + count_node(node.right)


def is_complete_bt(node, num_nodes, index=0):  # Default index values is 0
    if node is None:
        return True
    if index >= num_nodes:
        return False
    return is_complete_bt(node.left, num_nodes, 2 * index + 1) and is_complete_bt(node.right, num_nodes, 2 * index + 2)


root = Node(10)
root.left = Node(8)
root.right = Node(11)
root.left.left = Node(7)
root.left.right = Node(9)

if is_full_bt(root):
    print("This is a Full BT!")
else:
    print("This is not a Full BT!")

tree_depth = calculate_depth(root)
if is_perfect_bt(root, tree_depth):
    print("This is a Perfect BT!")
else:
    print("This is not a Perfect BT!")

num_of_nodes = count_node(root)
if is_complete_bt(root, num_of_nodes):
    print("This is a Complete BT!")
else:
    print("This is not a Complete BT!")

File: BinaryTree/JAVA/src/implementing_BT/BinaryTree.java
package implementing_BT;

//Full Binary Tree
//Perfect Binary Tree
//Complete Binary Tree
public class BinaryTree {
	Node root;

	// Check if this is a Full BT
	public boolean isFullBT(Node node) {
		if (node == null)
			return true;
		if (node.left == null && node.right == null)
			return true;
		if (node.left != null && node.right != null)
			return isFullBT(node.left) && isFullBT(node.right);

		return false;
	}

	public int calculateDepth(Node node) {
		int d = -1;
		while (node != null) {
			d++;
			node = node.left;
		}
		return d;
	}

	// Check if this is a Perfect BT
	public boolean isPerfectBT(Node node, int depth, int level) {
		if (node == null)
			return true;
		if (node.left == null && node.right == null)
			return depth == level + 1;
		if (node.left == null || node.right == null)
			return false;

		return isPerfectBT(node.left, depth, level + 1) && isPerfectBT(node.right, depth, level + 1);
	}

	public int countNode(Node node) {
		if (node == null)
			return 0;
		return 1 + countNode(node.left) + countNode(node.right);
	}

	// Check if this is a Complete BT
	public boolean isCompleteBT(Node node, int i, int numNodes) {
		if (node == null)
			return true;
		if (i >= numNodes)
			return false;

		return isCompleteBT(node.left, (2 * i) + 1, numNodes) && isCompleteBT(node.right, (2 * i) + 2, numNodes);
	}

	public static void main(String[] args) {
		BinaryTree myTree = new BinaryTree();
		myTree.root = new Node(10);
		myTree.root.left = new Node(8);
		myTree.root.right = new Node(11);
		myTree.root.left.left = new Node(7);
		myTree.root.left.right = new Node(9);

		if (myTree.isFullBT(myTree.root))
			System.out.println("This is a Full BT!");
		else
			System.out.println("This is not a Full BT!");

		int treeDepth = myTree.calculateDepth(myTree.root);
		if (myTree.isPerfectBT(myTree.root, treeDepth, 0))
			System.out.println("This is a Perfect BT!");
		else
			System.out.println("This is not a Perfect BT!");

		int numNodes = myTree.countNode(myTree.root);
		if (myTree.isCompleteBT(myTree.root, 0, numNodes))
			System.out.println("This is a Complete BT!");
		else
			System.out.println("This is not a Complete BT!");
	}
}